onnx/defs/nn/defs.cc

```diff
@@ -2487,4 +2486,211 @@ ONNX_OPERATOR_SET_SCHEMA(
+  // Original image has shape either [N, C, H, W] or [C, H, W]
+  //  but N and C are ignored for pads, dilations and stride computation
+  size_t n_input_dims = 2;
+
+  std::vector<int64_t> pads;
+  if (getRepeatedAttribute(ctx, "pads", pads)) {
+    if ((pads.size() != 0) && (pads.size() != n_input_dims * 2)) {
+      fail_shape_inference("Attribute pads has incorrect size");
+    }
+  }
+
+  std::vector<int64_t> dilations;
+  if (getRepeatedAttribute(ctx, "dilations", dilations)) {
+    if ((dilations.size() != 0) && (dilations.size() != n_input_dims)) {
+      fail_shape_inference("Attribute dilations has incorrect size");
+    }
+  }
+
+  std::vector<int64_t> strides;
+  if (getRepeatedAttribute(ctx, "strides", strides)) {
+    if ((strides.size() != 0) && (strides.size() != n_input_dims)) {
+      fail_shape_inference("Attribute strides has incorrect size");
+    }
+  }
+
+  // Validating inputs - all inputs are required
+  if (!hasNInputShapes(ctx, 3)) {
+    return;
+  }
+
+  // validating 'block_shape' input
+  std::vector<int64_t> block_shape;
+  const TensorProto* block_shape_data = ctx.getInputData(2);
+  if (block_shape_data) {
+    block_shape = ParseData<int64_t>(block_shape_data);
+    if (block_shape.size() != n_input_dims) {
+      fail_shape_inference("block_shape tensor must have 2 dimensions.");
+    }
+  }
+  if (ctx.getInputType(2)->tensor_type().shape().dim_size() != 1) {
+    fail_shape_inference("block_shape tensor must have rank 1.");
+  } else if (
+      (ctx.getInputType(2)->tensor_type().shape().dim(0).has_dim_value()) &&
+      (ctx.getInputType(2)->tensor_type().shape().dim(0).dim_value() != 2)) {
+    fail_shape_inference("block_shape tensor must have 2 dimensions.");
+  }
+
+  // validating 'input' input
+  auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+  if ((input_shape.dim_size() < 2) || (input_shape.dim_size() > 3)) {
+    fail_shape_inference("input must have rank 2 or 3.");
+  }
+
+  // validating 'image_shape' input
+  std::vector<int64_t> image_shape;
+  const TensorProto* image_shape_data = ctx.getInputData(1);
+  if (image_shape_data) {
+    image_shape = ParseData<int64_t>(image_shape_data);
+    if (image_shape.size() != 2) {
+      fail_shape_inference("image_shape tensor must have 2 dimensions.");
+    }
+  }
+  if (ctx.getInputType(1)->tensor_type().shape().dim_size() != 1) {
+    fail_shape_inference("image_shape tensor must have rank 1.");
+  } else if (
+      (ctx.getInputType(1)->tensor_type().shape().dim(0).has_dim_value()) &&
+      (ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value() != 2)) {
+    fail_shape_inference("image_shape tensor must have 2 dimensions.");
+  }
+
+  // inferencing final shape
+  int block_shape_size = 0;
+  if (static_cast<int>(block_shape.size()) > 0) {
+    block_shape_size = block_shape[0] * block_shape[1];
+  }
+  auto size_of_output = static_cast<int>(image_shape.size());
+  if ((size_of_output > 0) && (block_shape_size > 0)) {
```

Comment:
I think the above condition (and its else branch) are not required. The conditions in line 2575/2577 etc. below already handle both-cases. Does that sound right? Thanks!